04_Scheduling/04_Locks.md

```diff
@@ -0,0 +1,205 @@
+
+But how to solve this issue? We need a real atomic operation that has to be used within the acquire loop. There are different solutions that we can adopt (for example use the `xchgl` instruction in the for loop, that is guaranteed to be atomic), but as usual we want to keep thing extremely simple, and in this case we can avail of the help of the compiler (gcc) that provides a whole set of atomic functions to be used by our kernel. 
+
+In this case the function that we need to use is: 
+
+```c
+bool __atomic_test_and_set (void *ptr, int memorder)
+```
+
+This function perform an atomic _test and set_ operation on the byte ptr, as for the memorder parameter we are going to use `__ATOMIC_ACQUIRE` (for more details about the memorder parameter consult the gcc manual). If the test and set operation is succesfull (the lock was false and then set to true) the function return true, otherwise it will return false. 
+
+So to ensure multi-cpu isolation we can just replace our acquire code with the following: 
+
+```c
+void acquire(spinlock_t *lock) {
+    while(__atomic_test_and_set(&lock->locked, __ATOMIC_ACQUIRE));
+}
+```
+
+Of course we have also an atomic function to release the lock: 
+
+```c
+bool __atomic_release (void *ptr, int memorder)
+```
+
+And we need to update the release function too, but in this case we are going to use a different memorder value: `__ATOMIC_RELEASE`: 
+
+```c
+void release(spinlock_t *lock) {
+    __atomic_release(&lock->locked, __ATOMIC_RELEASE);
+}
+```
+
+Now we have a basic locking mechanism to protect shared resources on our systems. 
+
+## Where to go from here?
+
+What we have seen in this chapter is a simple locking mechanism to protect shared resources, as you can imagine this is not the only implementation, and it still has it's own limitation. For example the algorithm outlined above has the following problem
+
+* It doesn't guarantee the order of the processes requiring access to it. So if processes A, B and C require access to the shared resource, they are not guaranteed to be granted the lock in the same order, and in some scenario this could be a problem
+* It also doesn't prevent a _deadlock_ to happen. A _deadlock_ is a scenario where there are two or more processes that are waiting to acquire a shared resource that is hold by the other one. For example if A is holding resource X and needs to acquire a lock on Y, while B is holding a lock on Y and needs to acquire a lock on X, if not handled correctly we will have A and B stuck in the acquire loop forever.
+
+
+Solution to the problems above (and others that can arise) is a research area and out of scope of this guide, but is good to know that there are many algorithms available that are target to solve or mitigate these issues, and could be a good idea to add few of them to our kernel. 
+
```

Comment:
Very nice! I would also mention semaphores, and locks that interact with the scheduler. You could have a lock that if it can't acquire, it yields to the scheduler, rather than spinning. This is less responsive, but saves cpu time.
I know you mentioned deadlocks, but there is a special kind of deadlock that can occur in kernels with interrupts. Say a lock is taken (like in the PMM), and then an interrupt occurs. Now maybe the interrupt handler needs to allocate from the PMM, we have a deadlock. The solution to this is to disable interrupts anytime a lock is taken that might be accessed from an interrupt.